File: 그래프/BOJ-4803.py
# BOJ-4803.py
# 트리

import sys
input = sys.stdin.readline

def dfs(prev, node, graph, visited):
    visited[node] = True # 방문처리
    for next_node in graph[node]:
        # 다음 노드가 이전 노드와 같다면 DFS 수행 x (무시)
        if next_node == prev:
            continue
        # 다음 노드 != 이전 노드, 이미 방문했다면 사이클이 생긴것 (트리가 아님)
        if visited[next_node]:
            return False
        # 다음 노드 != 이전 노드, 다음 노드 첫 방문 -> DFS 수행
        # DFS 수행 도중 사이클이 생긴 경우에는 False 반환
        if not dfs(node, next_node, graph, visited):
            return False
    # DFS 수행 중 사이클이 생기지 않은 경우 True 반환
    return True

i = 1
n, m = map(int, input().split())
while n + m != 0:
    graph = [[] for _ in range(n+1)] # 인접 노드 저장하는 그래프
    visited = [False] * (n+1) # 방문 여부 확인하는 배열
    for _ in range(m):
        a, b = map(int, input().split())
        # 양방향으로 노드 저장
        graph[a].append(b) 
        graph[b].append(a)

    t = 0 # 트리의 개수
    for v in range(1, n+1):
        if not visited[v]: # 방문하지 않은 경우만 DFS 수행
            if dfs(0, v, graph, visited):
                t += 1 # DFS 수행결과 사이클이 없는 경우 트리의 개수 증가
    if t == 0:
        print("Case {}: No trees.".format(i))
    elif t == 1:
        print("Case {}: There is one tree.".format(i))
    else:
        print("Case {}: A forest of {} trees.".format(i, t))

    n, m = map(int, input().split())
    i += 1

File: 그래프/BOJ-4803.swift
//
//  main.swift
//  Algorithm
//
//  Created by Dain Song on 2021/05/28.
//

func DFS(prev: Int, curr: Int, graph: [[Int]], visited: inout [Bool]) -> Bool {
    
    visited[curr] = true // 방문처리
    
    for next in graph[curr] {
        // 다음과 이전이 같은 경우, 같은 경로를 뜻하므로 무시
        if next == prev {
            continue
        }
        // next != prev, 이미 방문한 경우 사이클이 생기므로 트리가 아님
        if visited[next] {
            return false
        }
        // next != prev, 사이클이 없는 경우 DFS 수행
        // DFS 수행 도중 사이클이 생기면 false가 반환됨 -> 재귀적으로 false 반환
        if !DFS(prev: curr, curr: next, graph: graph, visited: &visited) {
            return false
        }
    }
    // DFS 수행 도중 사이클이 생기지 않으면 true 반환.
    return true
}

func BOJ_4803() {
    var i = 0
    while true {
        let line = readLine()!.split(separator: " ").map({Int($0)!})
        let n = line[0], m = line[1]
        if n == 0 && m == 0 { break }
        i += 1

        // 그래프와 방문처리 배열 초기화 및 데이터 입력 처리
        var graph: [[Int]] = Array(repeating: Array<Int>(), count: n+1)
        var visited: [Bool] = Array(repeating: false, count: n+1)
        for _ in 0..<m {
            let line = readLine()!.split(separator: " ").map({Int($0)!})
            graph[line[0]].append(line[1])
            graph[line[1]].append(line[0])
        }
        
        var tree = 0 // 트리의 개수
        for node in 1...n {
            if !visited[node] {
                // 해당 노드를 방문하지 않은 경우에만 DFS 수행
                if DFS(prev: 0, curr: node, graph: graph, visited: &visited) {
                    tree += 1 // 사이클이 생기지 않은 경우 트리의 개수 증가
                }
            }
        }
        if tree == 0 {
            print("Case \(i): No trees.")
        } else if tree == 1 {
            print("Case \(i): There is one tree.")
        } else{
            print("Case \(i): A forest of \(tree) trees.")
        }
    }
}

BOJ_4803()


/* Input
6 3
1 2
2 3
3 4
6 5
1 2
2 3
3 4
4 5
5 6
6 6
1 2
2 3
1 3
4 5
5 6
6 4
0 0
*/

/* Output
Case 1: A forest of 3 trees.
Case 2: There is one tree.
Case 3: No trees.
 */
